File: interface/src/shared/api/memory.ts
import type {
  JsonValue,
  MemoryFact,
  MemoryEvent,
  MemoryProcedure,
  MemorySnapshot,
  MemoryStats,
} from "../types";
import { apiFetch } from "./core";

const memoryPath = (agentId: string, suffix = "") =>
  `/api/harness/agents/${encodeURIComponent(agentId)}/memory${suffix}`;

export const memoryApi = {
  // Facts
  listFacts: (agentId: string) =>
    apiFetch<MemoryFact[]>(memoryPath(agentId, "/facts")),
  getFact: (agentId: string, factId: string) =>
    apiFetch<MemoryFact>(memoryPath(agentId, `/facts/${factId}`)),
  getFactByKey: (agentId: string, key: string) =>
    apiFetch<MemoryFact>(memoryPath(agentId, `/facts/by-key/${key}`)),
  createFact: (agentId: string, data: { key: string; value: JsonValue; confidence?: number; importance?: number }) =>
    apiFetch<MemoryFact>(memoryPath(agentId, "/facts"), { method: "POST", body: JSON.stringify(data) }),
  updateFact: (agentId: string, factId: string, data: { value?: JsonValue; confidence?: number; importance?: number }) =>
    apiFetch<MemoryFact>(memoryPath(agentId, `/facts/${factId}`), { method: "PUT", body: JSON.stringify(data) }),
  deleteFact: (agentId: string, factId: string) =>
    apiFetch<void>(memoryPath(agentId, `/facts/${factId}`), { method: "DELETE" }),

  // Events
  listEvents: (agentId: string, params?: { limit?: number; since?: string; event_type?: string }) => {
    const query = new URLSearchParams();
    if (params?.limit) query.set("limit", String(params.limit));
    if (params?.since) query.set("since", params.since);
    if (params?.event_type) query.set("event_type", params.event_type);
    const qs = query.toString();
    return apiFetch<MemoryEvent[]>(memoryPath(agentId, `/events${qs ? `?${qs}` : ""}`));
  },
  createEvent: (agentId: string, data: { event_type: string; summary: string; metadata?: JsonValue; importance?: number }) =>
    apiFetch<MemoryEvent>(memoryPath(agentId, "/events"), { method: "POST", body: JSON.stringify(data) }),
  deleteEvent: (agentId: string, eventId: string) =>
    apiFetch<void>(memoryPath(agentId, `/events/${eventId}`), { method: "DELETE" }),

  // Procedures
  listProcedures: (agentId: string, params?: { skill?: string; min_relevance?: number }) => {
    const query = new URLSearchParams();
    if (params?.skill) query.set("skill", params.skill);
    if (params?.min_relevance != null) query.set("min_relevance", String(params.min_relevance));
    const qs = query.toString();
    return apiFetch<MemoryProcedure[]>(memoryPath(agentId, `/procedures${qs ? `?${qs}` : ""}`));
  },
  createProcedure: (agentId: string, data: {
    name: string; trigger: string; steps: string[];
    context_constraints?: JsonValue; skill_name?: string; skill_relevance?: number;
  }) =>
    apiFetch<MemoryProcedure>(memoryPath(agentId, "/procedures"), {
      method: "POST", body: JSON.stringify(data),
    }),
  updateProcedure: (agentId: string, procId: string, data: {
    name?: string; trigger?: string; steps?: string[];
    context_constraints?: JsonValue; skill_name?: string | null;
    skill_relevance?: number | null; success_rate?: number;
  }) =>
    apiFetch<MemoryProcedure>(memoryPath(agentId, `/procedures/${procId}`), {
      method: "PUT", body: JSON.stringify(data),
    }),
  deleteProcedure: (agentId: string, procId: string) =>
    apiFetch<void>(memoryPath(agentId, `/procedures/${procId}`), { method: "DELETE" }),

  // Aggregate
  getSnapshot: (agentId: string) =>
    apiFetch<MemorySnapshot>(memoryPath(agentId)),
  getStats: (agentId: string) =>
    apiFetch<MemoryStats>(memoryPath(agentId, "/stats")),
  wipeMemory: (agentId: string) =>
    apiFetch<void>(memoryPath(agentId), { method: "DELETE" }),
  triggerConsolidation: (agentId: string) =>
    apiFetch<void>(memoryPath(agentId, "/consolidate"), { method: "POST" }),
};

File: apps/aura-os-server/src/handlers/harness_proxy/memory.rs
use axum::extract::{Path, RawQuery, State};
use axum::http::Method;
use axum::response::Response;

use aura_os_core::AgentId;

use crate::state::AppState;

pub(crate) async fn list_facts(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory/facts"),
            query,
            None,
        )
        .await
}

pub(crate) async fn get_fact(
    State(state): State<AppState>,
    Path((agent_id, fact_id)): Path<(AgentId, String)>,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory/facts/{fact_id}"),
            None,
            None,
        )
        .await
}

pub(crate) async fn create_fact(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    body: String,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::POST,
            &format!("api/agents/{agent_id}/memory/facts"),
            None,
            Some(body),
        )
        .await
}

pub(crate) async fn update_fact(
    State(state): State<AppState>,
    Path((agent_id, fact_id)): Path<(AgentId, String)>,
    body: String,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::PUT,
            &format!("api/agents/{agent_id}/memory/facts/{fact_id}"),
            None,
            Some(body),
        )
        .await
}

pub(crate) async fn delete_fact(
    State(state): State<AppState>,
    Path((agent_id, fact_id)): Path<(AgentId, String)>,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::DELETE,
            &format!("api/agents/{agent_id}/memory/facts/{fact_id}"),
            None,
            None,
        )
        .await
}

pub(crate) async fn get_fact_by_key(
    State(state): State<AppState>,
    Path((agent_id, key)): Path<(AgentId, String)>,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory/facts/by-key/{key}"),
            None,
            None,
        )
        .await
}

pub(crate) async fn list_events(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory/events"),
            query,
            None,
        )
        .await
}

pub(crate) async fn create_event(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    body: String,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::POST,
            &format!("api/agents/{agent_id}/memory/events"),
            None,
            Some(body),
        )
        .await
}

pub(crate) async fn delete_event(
    State(state): State<AppState>,
    Path((agent_id, event_id)): Path<(AgentId, String)>,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::DELETE,
            &format!("api/agents/{agent_id}/memory/events/{event_id}"),
            None,
            None,
        )
        .await
}

pub(crate) async fn list_procedures(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory/procedures"),
            query,
            None,
        )
        .await
}

pub(crate) async fn get_procedure(
    State(state): State<AppState>,
    Path((agent_id, proc_id)): Path<(AgentId, String)>,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory/procedures/{proc_id}"),
            None,
            None,
        )
        .await
}

pub(crate) async fn create_procedure(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    body: String,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::POST,
            &format!("api/agents/{agent_id}/memory/procedures"),
            None,
            Some(body),
        )
        .await
}

pub(crate) async fn update_procedure(
    State(state): State<AppState>,
    Path((agent_id, proc_id)): Path<(AgentId, String)>,
    body: String,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::PUT,
            &format!("api/agents/{agent_id}/memory/procedures/{proc_id}"),
            None,
            Some(body),
        )
        .await
}

pub(crate) async fn delete_procedure(
    State(state): State<AppState>,
    Path((agent_id, proc_id)): Path<(AgentId, String)>,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::DELETE,
            &format!("api/agents/{agent_id}/memory/procedures/{proc_id}"),
            None,
            None,
        )
        .await
}

pub(crate) async fn list_procedures_by_skill(
    State(state): State<AppState>,
    Path((agent_id, skill_name)): Path<(AgentId, String)>,
    RawQuery(query): RawQuery,
) -> Result<Response, axum::http::StatusCode> {
    let mut qs = format!("skill={skill_name}");
    if let Some(q) = query {
        qs = format!("{qs}&{q}");
    }
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory/procedures"),
            Some(qs),
            None,
        )
        .await
}

pub(crate) async fn get_memory_snapshot(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory"),
            query,
            None,
        )
        .await
}

pub(crate) async fn wipe_memory(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::DELETE,
            &format!("api/agents/{agent_id}/memory"),
            None,
            None,
        )
        .await
}

pub(crate) async fn get_memory_stats(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    RawQuery(query): RawQuery,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::GET,
            &format!("api/agents/{agent_id}/memory/stats"),
            query,
            None,
        )
        .await
}

pub(crate) async fn trigger_consolidation(
    State(state): State<AppState>,
    Path(agent_id): Path<AgentId>,
    body: String,
) -> Result<Response, axum::http::StatusCode> {
    state
        .harness_http
        .proxy_json(
            Method::POST,
            &format!("api/agents/{agent_id}/memory/consolidate"),
            None,
            Some(body),
        )
        .await
}
